drivers: sdhc: add sdhc driver for Ambiq MCU

darrenlu-ambiq opened the pull request:
Add Ambiq SDHC/SDIO driver support for apollo4p_evb.

AlessandroLuo commented:
@danieldegrasse please help to review when you get time, thank you

danieldegrasse commented on drivers/sdhc/sdhc_ambiq.c:
```diff
@@ -0,0 +1,829 @@
+{
+	LOG_DBG("SDHC get host props");
+	memset(props, 0, sizeof(*props));
+	props->f_max = AMBIQ_SDIO_FREQ_MAX;
```
Are these properties of the IP itself? Might make sense to place them in DTS- I know in many cases that various instantiations of the same IP have different upper frequency limits

darrenlu-ambiq replied:
Done. Moved to DTS.

danieldegrasse commented on drivers/sdhc/sdhc_ambiq.c:
```diff
@@ -0,0 +1,829 @@
+	memset(props, 0, sizeof(*props));
+	props->f_max = AMBIQ_SDIO_FREQ_MAX;
+	props->f_min = AMBIQ_SDIO_FREQ_MIN;
+	props->power_delay = 50; /* power_delay could be changed based on board power design */
```
If this can be changed based on board, make it a devicetree property please :)

darrenlu-ambiq replied:
Done. Moved to DTS.

danieldegrasse commented on drivers/sdhc/sdhc_ambiq.c:
```diff
@@ -0,0 +1,829 @@
+	props->host_caps.sdr50_support = true;
+	props->host_caps.sdr104_support = true;
+	props->host_caps.ddr50_support = true;
+	props->host_caps.hs200_support = true; /* Ambiq SDIO only supports 96MHz MAX frequency */
```
Ok- just to be clear here, this IP supports SDR104/HS200, but the maximum frequency is 96MHz? That is possible, I just want to make sure it is accurate since the SD host stack relies on these fields to determine if the card should enter UHS mode or not

darrenlu-ambiq replied:
Yes. the IP supports SDR104/HS200. But max frequency is limited to 96MHz in Apollo4P silicon.

danieldegrasse commented on drivers/sdhc/sdhc_ambiq.c:
```diff
@@ -0,0 +1,829 @@
+	 * Recommend to use 1.8V IO for SDIO Host controller.
+	 * External level shifter is required for voltage switch between 1.8V and 3V.
+	 */
+	if (ios->signal_voltage != SD_VOL_1_8_V) {
```
Why is there a switch statement above if we are going to reject signal voltages besides 1.8V? Also, this won't work with the SD stack- cards start at 3.3V, and negotiate down to 1.8V

danieldegrasse commented on drivers/sdhc/sdhc_ambiq.c:
```diff
@@ -0,0 +1,829 @@
+	config->irq_config_func(dev);
+
+	/* Check whether card is present */
+	while (am_hal_card_host_find_card(data->host, &data->card) != AM_HAL_STATUS_SUCCESS) {
```
The SDHC init should not be where we check card presence- the controller can (and should) come up with or without a card present. The SD stack is responsible for detecting if a card is present

darrenlu-ambiq replied:
Both SDHC host init and card init are done in Ambiq SDHC HAL. This is needed to match with Ambiq SDHC HAL.

danieldegrasse replied:
The card init is done in the HAL? So essentially the driver initializes the card, and then relies on the MMC stack to reset the card and reinitialize it? If so, we need to find a way to *just* initialize the host within the HAL- the HAL should not handle card init logic.

darrenlu-ambiq replied:
The card init is done in HAL API am_hal_card_init() which is not called in zephyr sdhc driver. The function am_hal_card_host_find_card() is mainly to check the EMMC presence before calling am_hal_card_init(). It also sets the default value to init the card structure. Since am_hal_card_init() is not needed for zephyr, so I just removed the code calling am_hal_card_host_find_card() and moved the card structure init to sdhc_ambiq driver.
 tests/subsys/sd/mmc and samples/subsys/fs/fs_sample are passed on apollo4p_evb after the change. Please help review again. Thank you.

danieldegrasse commented on dts/bindings/sdhc/ambiq,sdhc.yaml:
Please put the binding file for the SDHC IP in the prior commit with the driver

danieldegrasse reviewed (commented):
Generally looks fine, looks like this is tested with the MMC stack?

A few requests:
- please at least make an attempt to support the SD stack here (don't reject voltage configurations other than 1.8V)
- Ideally (if a board exists that supports it) add a DTS definition that will test the SD stack with this controller
- Add the relevant overlays to enable this board with `tests/subsys/sd/mmc`

danieldegrasse commented:
> @danieldegrasse please help to review when you get time, thank you

Sorry for the delay getting to this- thank you for the ping

darrenlu-ambiq committed:
drivers: sdhc: add sdhc driver for Ambiq MCU

Add SDHC/SDIO driver support for Ambiq Apollo4 MCU.

Signed-off-by: Darren Lu <dlu@ambiq.com>

darrenlu-ambiq committed:
dts: add SDIO support to apollo4p_evb

Update dts for using SDIO on apollo4p_evb.

Signed-off-by: Darren Lu <dlu@ambiq.com>

darrenlu-ambiq committed:
samples: add apollo4p_evb to fs_sample

Add apollo4p_evb.conf to fs_sample.

Signed-off-by: Darren Lu <dlu@ambiq.com>

darrenlu-ambiq committed:
west.yml: update hal_ambiq to support sdhc

Updated Apollo4P HAL to support SDHC driver.

Signed-off-by: Darren Lu <dlu@ambiq.com>

darrenlu-ambiq commented:
> Generally looks fine, looks like this is tested with the MMC stack?
> 
> A few requests:
> 
> * please at least make an attempt to support the SD stack here (don't reject voltage configurations other than 1.8V)
> * Ideally (if a board exists that supports it) add a DTS definition that will test the SD stack with this controller
> * Add the relevant overlays to enable this board with `tests/subsys/sd/mmc`

@danieldegrasse Thank you for the review. The SDHC driver is tested with MMC stack on apollo4p_evb. tests/subsys/sd/mmc and samples/subsys/fs/fs_sample are passed. Following are the pass log snapshots. Apollo4P only support 1.8V IO so that external 3.3V  level shifter is required for SD card. apollo4p_evb does not support SD card which is not tested. I have removed the code which rejects voltage configuration. SDHC is enabled in apollo4p_evb.dts by default. I have moved max & min frequency and power_delay settings into device tree. Please help review again.

![image](https://github.com/user-attachments/assets/1bdce352-0bd6-4912-bd59-0456db7597cd)
![image](https://github.com/user-attachments/assets/be91654d-4543-491d-a27d-d6d97caccb0a)
